File: ui/lib/tournament_ui/workflow_runs.ex
defmodule TournamentUi.WorkflowRuns do
  @moduledoc """
  Read-only Elixir adapter over the Python-side `workflow_run` table — the
  queryable projection of Temporal release-workflow state (ADR 0001 §4 step
  6, see bin/workflow_runs.py).

  Temporal is the source of truth; rows are written ONLY by Python Temporal
  Activities. Elixir reads and never executes DDL. Older DBs may predate the
  table entirely — every reader treats "no such table" as empty, never a
  crash (Catalog precedent).
  """

  alias Exqlite.Sqlite3

  @columns "id, spec_digest, temporal_workflow_id, temporal_run_id, status, " <>
             "environment_id, detail, stage_history, started_at, finished_at"

  @doc """
  Workflow runs, newest first (id DESC), optionally filtered by status.

  Options: `status:` (string), `limit:` (default 50). Each row carries a
  `stage_count` derived from the parsed `stage_history`.
  """
  def list_runs(opts \\ []) do
    status = Keyword.get(opts, :status)
    limit = Keyword.get(opts, :limit, 50)

    {where, params} =
      case status do
        nil -> {"", []}
        s -> {" WHERE status = ?", [s]}
      end

    sql =
      "SELECT #{@columns} FROM workflow_run#{where} ORDER BY id DESC LIMIT ?"

    case query(sql, params ++ [limit]) do
      {:ok, rows} -> Enum.map(rows, &row_to_map/1)
      _ -> []
    end
  end

  @doc """
  Newest run for a temporal workflow id (retries/continue-as-new mint new
  run ids), with parsed `detail` and `stage_history`. Nil when unknown or
  when the table is missing.
  """
  def get_run(workflow_id) when is_binary(workflow_id) do
    sql =
      "SELECT #{@columns} FROM workflow_run WHERE temporal_workflow_id = ? " <>
        "ORDER BY id DESC LIMIT 1"

    case query(sql, [workflow_id]) do
      {:ok, [row | _]} -> row_to_map(row)
      _ -> nil
    end
  end

  def get_run(_), do: nil

  # ── private ────────────────────────────────────────────────────────────

  defp row_to_map([
         id,
         spec_digest,
         workflow_id,
         run_id,
         status,
         environment_id,
         detail,
         stage_history,
         started_at,
         finished_at
       ]) do
    stages = decode_list(stage_history)

    %{
      id: id,
      spec_digest: spec_digest,
      temporal_workflow_id: workflow_id,
      temporal_run_id: run_id,
      status: status,
      environment_id: environment_id,
      detail: decode_map(detail),
      stage_history: stages,
      stage_count: length(stages),
      started_at: started_at,
      finished_at: finished_at
    }
  end

  defp decode_map(s) when is_binary(s) do
    case Jason.decode(s) do
      {:ok, m} when is_map(m) -> m
      _ -> %{}
    end
  end

  defp decode_map(_), do: %{}

  defp decode_list(s) when is_binary(s) do
    case Jason.decode(s) do
      {:ok, l} when is_list(l) -> l
      _ -> []
    end
  end

  defp decode_list(_), do: []

  defp db_path do
    home = System.get_env("DATA_TOURNAMENTS_HOME") || "/tmp/data-tournaments"
    Path.join(home, "judgements.db")
  end

  # Catalog.query/2 shape: readonly open, busy_timeout, missing tables raise
  # inside prepare, get caught, and callers translate to empty results.
  defp query(sql, params) do
    if not File.exists?(db_path()) do
      {:error, :no_db}
    else
      {:ok, conn} = Sqlite3.open(db_path(), mode: :readonly)

      try do
        :ok = Sqlite3.execute(conn, "PRAGMA busy_timeout = 5000")
        {:ok, stmt} = Sqlite3.prepare(conn, sql)
        :ok = Sqlite3.bind(stmt, params)
        rows = collect(conn, stmt, [])
        Sqlite3.release(conn, stmt)
        {:ok, Enum.reverse(rows)}
      catch
        kind, value -> {:error, {kind, value}}
      after
        Sqlite3.close(conn)
      end
    end
  end

  defp collect(conn, stmt, acc) do
    case Sqlite3.step(conn, stmt) do
      {:row, row} -> collect(conn, stmt, [row | acc])
      :done -> acc
    end
  end
end

File: bin/workflow_runs.py
"""workflow_run: queryable projection of Temporal release-workflow state.

ADR 0001 §4 step 6. Temporal is the SOURCE OF TRUTH for execution state;
this table exists so the LiveView audit UI can list/inspect runs with plain
SQL joins. Written ONLY by Python Temporal Activities (never by workflow
bodies — those must stay deterministic — and never by Elixir), mirroring the
optimizer_run precedent (bin/optimizer_runs.py).

If this projection is ever lost or corrupted it can be rebuilt from
Temporal's own history via the frontend API; nothing here is authoritative.
"""
from __future__ import annotations

import json
import os
import sqlite3
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

TERMINAL_STATUSES = ("done", "failed", "canceled", "rolled-back")
STATUSES = ("running", "awaiting-approval") + TERMINAL_STATUSES

def _db_path() -> Path:
    home = Path(os.environ.get("DATA_TOURNAMENTS_HOME", "/tmp/data-tournaments"))
    return home / "judgements.db"

def _connect() -> sqlite3.Connection:
    conn = sqlite3.connect(_db_path())
    conn.row_factory = sqlite3.Row
    conn.execute("PRAGMA foreign_keys = ON")
    conn.execute("PRAGMA busy_timeout = 5000")
    return conn

def _now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")

def init() -> None:
    """Ensure the schema (incl. workflow_run) exists. Idempotent."""
    from bin import catalog

    catalog.init()

def start(
    *,
    temporal_workflow_id: str,
    temporal_run_id: str,
    spec_digest: Optional[str] = None,
    environment_id: Optional[int] = None,
    detail: Optional[dict[str, Any]] = None,
) -> int:
    """Record a new run. Idempotent on (workflow_id, run_id): re-recording
    the same Temporal execution returns the existing row id — Activities
    retry, and a retried record-start must not mint duplicate rows."""
    init()
    with _connect() as conn:
        existing = conn.execute(
            "SELECT id FROM workflow_run "
            "WHERE temporal_workflow_id=? AND temporal_run_id=?",
            (temporal_workflow_id, temporal_run_id),
        ).fetchone()
        if existing:
            return int(existing["id"])
        cur = conn.execute(
            "INSERT INTO workflow_run(spec_digest, temporal_workflow_id, "
            "temporal_run_id, environment_id, detail) VALUES (?, ?, ?, ?, ?)",
            (
                spec_digest,
                temporal_workflow_id,
                temporal_run_id,
                environment_id,
                json.dumps(detail or {}),
            ),
        )
        conn.commit()
        return int(cur.lastrowid)

def record_stage(
    run_id: int, *, stage: str, status: str, detail: Optional[dict[str, Any]] = None
) -> None:
    """Append one entry to the run's append-only stage history.

    History is never rewritten: each call appends {stage, status, at}. The
    run's coarse status is NOT changed here — use set_status for that.
    """
    entry = {"stage": stage, "status": status, "at": _now()}
    if detail:
        entry["detail"] = detail
    with _connect() as conn:
        row = conn.execute(
            "SELECT stage_history FROM workflow_run WHERE id=?", (run_id,)
        ).fetchone()
        if row is None:
            raise KeyError(f"workflow_run id {run_id} not found")
        history = json.loads(row["stage_history"])
        history.append(entry)
        conn.execute(
            "UPDATE workflow_run SET stage_history=? WHERE id=?",
            (json.dumps(history), run_id),
        )
        conn.commit()

def set_status(
    run_id: int, status: str, *, detail: Optional[dict[str, Any]] = None
) -> None:
    """Set the coarse run status; stamps finished_at on terminal statuses.

    Terminal statuses are sticky: a projection update arriving late (retried
    activity) cannot flip done/failed/rolled-back back to running — mirrors
    the judgement.py done-flip guard.
    """
    if status not in STATUSES:
        raise ValueError(f"unknown status {status!r}; expected one of {STATUSES}")
    with _connect() as conn:
        row = conn.execute(
            "SELECT status, detail FROM workflow_run WHERE id=?", (run_id,)
        ).fetchone()
        if row is None:
            raise KeyError(f"workflow_run id {run_id} not found")
        if row["status"] in TERMINAL_STATUSES:
            return
        merged = json.loads(row["detail"])
        if detail:
            merged.update(detail)
        if status in TERMINAL_STATUSES:
            conn.execute(
                "UPDATE workflow_run SET status=?, detail=?, finished_at=? "
                "WHERE id=?",
                (status, json.dumps(merged), _now(), run_id),
            )
        else:
            conn.execute(
                "UPDATE workflow_run SET status=?, detail=? WHERE id=?",
                (status, json.dumps(merged), run_id),
            )
        conn.commit()

def get(run_id: int) -> Optional[dict[str, Any]]:
    with _connect() as conn:
        row = conn.execute(
            "SELECT * FROM workflow_run WHERE id=?", (run_id,)
        ).fetchone()
    return _to_dict(row) if row else None

def get_by_workflow_id(temporal_workflow_id: str) -> list[dict[str, Any]]:
    """All runs for a workflow id (retries/continue-as-new mint new run ids),
    newest first."""
    with _connect() as conn:
        rows = conn.execute(
            "SELECT * FROM workflow_run WHERE temporal_workflow_id=? "
            "ORDER BY id DESC",
            (temporal_workflow_id,),
        ).fetchall()
    return [_to_dict(r) for r in rows]

def list_runs(
    *, status: Optional[str] = None, limit: int = 50
) -> list[dict[str, Any]]:
    q = "SELECT * FROM workflow_run"
    args: tuple = ()
    if status:
        q += " WHERE status=?"
        args = (status,)
    q += " ORDER BY id DESC LIMIT ?"
    args += (int(limit),)
    with _connect() as conn:
        rows = conn.execute(q, args).fetchall()
    return [_to_dict(r) for r in rows]

def _to_dict(row: sqlite3.Row) -> dict[str, Any]:
    d = dict(row)
    d["detail"] = json.loads(d["detail"])
    d["stage_history"] = json.loads(d["stage_history"])
    return d
